docs/topic/infrastructure/cluster-design.md

```diff
@@ -89,6 +91,28 @@ that `prometheus-server` may require more memory than is available.
 On EKS we always use the `r5.xlarge` nodes to avoid running low on allocatable
 pods.
 
+#### For notebook node pool
+
+We default to always having available three machine types of 4 / 16 / 64 CPU and a memory specification of 32 / 128 / 512 GB for each notebook node poll in a 2i2c cluster. These three options have proven to be general enough to cover most usage scenarios, including events as well as being a good trade off between available options and the maintainability toil.
+
+```{note}
+The actual resulting capacity in the k8s cluster is slightly lower and dependent of the cloud provider.
+```
+
+The three machine types based on the cloud provider are the following:
+- [GKE](https://cloud.google.com/compute/docs/general-purpose-machines)
+  - n2-highmem-4
+  - n2-highmem-16
+  - n2-highmem-64
+- [EKS](https://aws.amazon.com/ec2/instance-types/r5/)
+  - r5.xlarge
```

Comment:
### Choice of 4 16 64 (set A) or 4 8 32 (set B)

Observations:

- Set A is _evenly distributed_ and has a _wider span_, while set B has higher resolution at the smaller sizes.
- Set A is the current default in our terraform templates and many clusters already have set A configured

"Based on existing data, we should pick smaller sizes." doesn't seem to follow when I think about it. I reason like this:

1. Set A and B alike have the 4 CPU / 32 GB node as the smallest size and can handle most resource allocation requests. Larger sizes is mostly something I see as a way to optimize a) startup times, and b) node utilization for CPU or memory resources _for which requests are lower than the limit_ (the [central limit theory](https://en.wikipedia.org/wiki/Central_limit_theorem)) then can enable a higher utilization with less risk of running out of the resource then).
2. A 4x multiplier between instance sizes seems sufficient for the resolution when choosing how many users we enable to schedule per node on average.
3. The very large 64 CPU node (and 32 CPU) are mostly going to be used during events:
   1. There have been events with ~8 GB or ~16 GB of memory is requested, a 64 node would fit 64 or 32 users, which could be reasonable "users per node" choices. 32 / 16 users per node is probably fine as well though in many situations.
   2. Anyone pre-starting a 64 node could pre-start / pre-image-pull for a double the users to minimize delays without having us pre-start + pre-image-pull manually.
   
I'm currently favoring set A the most, and favoring set C of 4 8 16 64 or over set B. Set C would a) retain the wider span of set A, b) provide even more resolution in the choice among the smaller instance sizes, and c) would still involve 16 CPU node which is actively used in a few clusters and the 64 node that is setup but not actively used in a few.

@yuvipanda @GeorgianaElena how do you rank the various instance type combinations we consider ensuring is available for all clusters?

Reply:
To make things clearer, I made a third row in https://grafana.pilot.2i2c.cloud/d/ed8d55b8-54c7-4658-bea0-f9659a9b7c33/global-resource-usage?orgId=1&var-PROMETHEUS_DS=All, which shows only nodes that have had 256GB+ memory usage even once. You'll see that it's completely empty over the last 30 days - we have had 0 nodes that have ever used more than 256GB of RAM. And one single cluster had used that in the last 90 days for one day over 3 nodes.

So yes, I'm advocating for much smaller instances, because our current instances are way too big and costing our users a lot of money.

Reply:
Ok, I made another row, now for nodes that have used over 128GiB. The *only* cluster where that has happened is the same cluster that had gone over 256GB of RAM for a day, and nowhere else (over 90 days). This makes me conclude that even 256GB is too big, and we should be *even smaller*. I propose the equivalent of:

- r5.xlarge
- r5.2xlarge
- r5.4xlarge
- r5.8xlarge (optionally present always, but no profileList actually deploys here by default. We can offer if needed. By being present by default, engineers don't have to touch eksctl / terraform when this request comes in).

Anything else should be considered 'large nodes' and be deployed on user requests, not provided by default.

Reply:
The equivalent of this in GCP sizes would be: 4 8 16 (32 optional)

Reply:
I also explored memory *requests* and so there's another row in grafana now of nodes with memory requests that exceeded 256GiB. 

Over the last 6 months, other than in two clusters that consistently have terabyte sized requests (probably JMTE and carbonplan with their x instances?), I count 6 (or 9) separate instances where there has existed a node with 256+ GB of memory requests. There's also one that had about 7T of memory requests that I'm sure is some kinda error.

I added another row with nodes that have more than 128GiB of requests. There is definitely more here, but it's very periodic and only has occured even once in 9 of our 28 clusters. To me, this continues to present a strong case for provisioning: 4 8 16, and 32 provisioned but not used by default (only used during events). I'm alright with also provisioning 64 but *not* using it by default, and only using it for events under specific conditions.

Reply:
So I guess my concrete proposal is:

## On Terraform / `eksctl`

Provisioning nodes here has no cost, and helps engineers make adjustments to profileList more easily with less toil. So it's ok to provision nodegroups here that aren't actively in use via profileList. So we provision:

On GCP
1. n2-highmem-4
2. n2-highmem-8
3. n2-highmem-16
4. n2-highmem-32

On AWS
1. r5.xlarge
2. r5.2xlarge
3. r5.4xlarge
4. r5.8xlarge

On Azure
(equivalent memory sizes, as their website is far too confusing)

## Default ProfileList

This is more important for actual usage and UX, and from what I can tell, this PR doesn't actually say *anything* about profileList at all. Which is perfectly fine! I can provide similar feedback whenever that is being worked on, and we needn't concern ourselves with that in this PR.

Reply:
I don't feel heard for what I've said @yuvipanda, maybe you don' either. I think we wont either if this discussion is continued async. Can we chat about this sync a bit?

Reply:
I'm sorry you don't feel heard, @consideRatio. I don't either :( I'll reach out on slack.

Reply:
We had a nice conversation on slack and https://github.com/2i2c-org/infrastructure/pull/3304#pullrequestreview-1694081796 is a good way forward. We can come back to https://github.com/2i2c-org/infrastructure/issues/3307 later.